kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,1180 @@
+                rEquationIds.reserve(master_equation_ids.size());
+                for (auto &master_eq_id : master_equation_ids)
+                {
+                    // Add the current slaves master eq ids to the equation ids
+                    rEquationIds.push_back(master_eq_id);
+                }
+            }
+        }
+        KRATOS_CATCH("ResidualBasedBlockBuilderAndSolverWithConstraints::ApplyConstraints failed ..");
+    }
+
+    /**
+     * @brief   This function modifies the LHS and RHS of the rCurrentContainer to account for any master-slave constraints its nodes/dofs 
+     *          are carrying.
+     * @details Here cannot use the pure Geometry because, we would need the dof list from the element/geometry.
+     * @param   rModelPart The model part of the problem to solve
+     * @param   rCurrentContainer the element or condition where the rEquationIds to be modified for master-slave constraints
+     * @param   rLHSContribution the LHS contibution of the rCurrentContainer
+     * @param   rRHSContribution the RHS contibution of the rCurrentContainer
+     * @param   rEquationIds the equation id vector for the above element or condition
+     * @param   rCurrentProcessInfo the current process info
+     */
+    template <typename TContainerType>
+    void ApplyConstraints(ModelPart& rModelPart,
+                          TContainerType& rCurrentContainer,
+                          LocalSystemMatrixType& rLHSContribution,
+                          LocalSystemVectorType& rRHSContribution,
+                          typename TContainerType::EquationIdVectorType& rEquationIds,
+                          ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY
+        // If no slave is found for this container , no need of going on
+        if (! HasSlaveNode(rCurrentContainer.GetGeometry()))
+            return;
+
+        typename TContainerType::EquationIdVectorType equation_ids = rEquationIds;
+        // Saving th original system size
+        const IndexType initial_sys_size = rLHSContribution.size1();
+
+        // Formulating which are the internal, slave indices locally.
+        LocalIndices local_indices;
+
+        // first fill in the rEquationIds using the above function (overloaded one)
+        ApplyConstraints<TContainerType>(rModelPart, rCurrentContainer, rEquationIds, rCurrentProcessInfo); // now rEquationIds has all the slave equation ids appended to it.
+        IndexType total_number_of_masters = rEquationIds.size() - initial_sys_size;
+        // Calculating the local indices corresponding to internal, master, slave dofs of this container
+        CalculateLocalIndices(rEquationIds, local_indices, total_number_of_masters);
+
+        // resizing the matrices to the new required length
+        MatrixType lhs_contribution = rLHSContribution;
```

Comment:
I made copies because at a later stage the algorithm needs the unmodified LHS and RHS contributions.  For example here : https://github.com/KratosMultiphysics/Kratos/blob/9b2df88f88d23c7df93c12de665ea25bb3c808c0/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h#L883

Reply:
ok